### ISSUE TITLE
False positive when using a TypedDict after multiple branches may assign to it: all types inside TypedDict turn into object

### ISSUE THREAD

**Issue by Wizzerinus** [microsoft/pyright#11743]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
When multiple code branches assign to a TypedDict-typed variable, Pyright incorrectly discards all information about the types inside the TypedDict and thinks they're all `object`s. This makes it impossible to use the dict variable in any way without a type cast.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from typing import TypedDict, Unpack, assert_type

class AwesomeDict(TypedDict, total=False):
    foo: int
    bar: int
    baz: int

def consume_a(arg: str, **kw: Unpack[AwesomeDict]): pass
def get_bool() -> bool: raise NotImplementedError

a: AwesomeDict = {}
if get_bool():
    a = {"foo": 2, "bar": 3}
else:
    a = {"foo": 3, "baz": 4}

assert_type(a, AwesomeDict)
# Expected: no error
# Actual:
#    "assert_type" mismatch: expected "AwesomeDict" but received "AwesomeDict"
consume_a("hi", **a)
# Expected: no error
# Actual:
#    Argument of type "object" cannot be assigned to parameter "foo" of type "int" in function "consume_a"
#      "object" is not assignable to "int"
#    Argument of type "object" cannot be assigned to parameter "bar" of type "int" in function "consume_a"
#      "object" is not assignable to "int"
#    Argument of type "object" cannot be assigned to parameter "baz" of type "int" in function "consume_a"
#      "object" is not assignable to "int"
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Pyright 1.1.412, Pyright Playground also reproduces this bug.
